crates/matrix-sdk-crypto/src/machine.rs

```diff
@@ -1397,60 +1397,39 @@ impl OlmMachine {
-                // `InboundGroupSession` or not.
+        let sender_data = SenderDataFinder::find_using_curve_key(
+            self.store(),
+            session.sender_key(),
+            sender,
+            session,
+        )
+        .await?;
 
-                let link_problem = if session.has_been_imported() {
+        Ok(match sender_data {
+            SenderData::UnknownDevice { owner_check_failed: false, .. } => {
+                let device_link_problem = if session.has_been_imported() {
                     DeviceLinkProblem::InsecureSource
                 } else {
                     DeviceLinkProblem::MissingDevice
                 };
 
-                (VerificationState::Unverified(VerificationLevel::None(link_problem)), None)
+                (VerificationState::Unverified(VerificationLevel::None(device_link_problem)), None)
             }
-            Some(device) => {
-                let device_id = device.device_id().to_owned();
-
-                // We found a matching device, let's check if it owns the session.
-                if !(device.is_owner_of_session(session)?) {
-                    // The key cannot be linked to an owning device.
-                    // Refuse to provide the device_id since this device is not the owner of this
-                    // session.
-                    (
-                        VerificationState::Unverified(VerificationLevel::None(
-                            DeviceLinkProblem::InsecureSource,
-                        )),
-                        None,
-                    )
-                } else {
-                    // We only consider cross trust and not local trust. If your own device is not
-                    // signed and send a message, it will be seen as Unverified.
-                    if device.is_cross_signed_by_owner() {
-                        // The device is cross signed by this owner Meaning that the user did self
-                        // verify it properly. Let's check if we trust the identity.
-                        if device.is_device_owner_verified() {
-                            (VerificationState::Verified, Some(device_id))
-                        } else {
-                            (
-                                VerificationState::Unverified(
-                                    VerificationLevel::UnverifiedIdentity,
-                                ),
-                                Some(device_id),
-                            )
-                        }
-                    } else {
-                        // The device owner hasn't self-verified its device.
-                        (
-                            VerificationState::Unverified(VerificationLevel::UnsignedDevice),
-                            Some(device_id),
-                        )
-                    }
-                }
+            SenderData::UnknownDevice { owner_check_failed: true, .. } => (
```

Comment:
Shame that we can't put this into a `From` implementation, would cement down the fact that these two are parts of the same coin.

Reply:
Yeah I considered that, and in a future PR I moved it into a function, but it's `(SenderData, is_imported) -> (VerificationState, DeviceId)` so it doesn't quite fit.

Reply:
Yeah, I saw that `is_imported` messes everything up :cry:.